### PR TITLE
[BottomNavigation] Adds a large item dialog view

### LINKED ISSUE #7122: [BottomNavigation] Create pop-up view for .largeContentSizeImage support
This was filed as an internal issue. If you are a Googler, please visit [b/130344111](http://b/130344111) for more details.

---
#### Internal data
- Associated internal bug: [b/130344111](http://b/130344111)

### PR DESCRIPTION
**Pull request by nokeeo**
Adds a view for displaying a large icon and title when the user has dynamic type enabled.

This is part of a larger effort to add dynamic type support to `MDCBottomNavigationBarController`.

Closes #7122

### CODE REVIEW

**Comment by jverkoey** [comment]
Because this is a part of a larger effort, please add context to the PR description on the need for a separate `MDCBottomNavigationLargeItemDialogView` + `MDCBottomNavigationSystemDialogView` implementation. Is it possible to consolidate these two types into one?

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h]: Tip: Please use `@c` in front of variable/property/method names.

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m]: Nit/naming: Please name this `commonMDCBottomNavigationSystemDialogViewInit` to match the general pattern of the code base.

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m:76]: Will the blur view not inherit the layoutMargins from its view hierarchy?  It would be nice if we could just set `self.layoutMargins`. If not, let's document why that seems to be the case.

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m]: Tip: Please consider setting the blurView bounds + center rather than the frame. Alternatively, consider relying on the autoresizing mask (I see you haven't disabled it).

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Please document this constant. Its purpose is not clear to me from reading the name.  Specifically, I'm not sure how to compare a height (scalar) to a dialog (object) to produce a ratio.

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h]: Please mark the `item` paramter as `nonnull`

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Please make this API `nonnull` if possible.

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Is this the behavior of UITabBarController? Since MDCBottomNavigationBar supports multiple lines I want to make sure we're intentionally removing that capability here.

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (romoore) [inline components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Tip: I'd prefer to see a fixed height for the image, consistent across devices.  If the view has additional vertical space, I think it's worth dedicating it to the label where possible.

Comment (romoore) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: This also doesn't seem to match the behavior of UITabBarController in my testing.  The image below shows a blown-up version with the vertical midpoint identified with the dotted line.

Screen Shot 2019-04-16 at 8 42 44 AM

**Review by romoore — CHANGES_REQUESTED** [review]
Please add snapshot tests covering the following cases:

1.  Background colors
    1.   Black
    1.   White
    1.   Red
    1.   Green
    1.   Blue
1.  Various view bounds sizes
    1.   Small view bounds (120 x 120)
    1.   Medium view bounds (240 x 240)
    1.   Large view bounds (480 x 480)

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h]: Done.

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m]: Done.

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m:76]: I believe layout margins only propagate to a view's subviews if the subview has `preservesSuperviewLayoutMargins` set to `YES`.  I am hesitant to use this approach, because a change in the view hierarchy of `UIVisualEffectView`  may break the margins functionality.  Thoughts?

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m]: Done.  Just curious why using bounds + center is preferred over setting the frame?

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Constant removed with latest update.

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.h]: Done.

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Done.  Made the `item` parameter `_Nonnull`, but the return value if nil by design.

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: If there is an image, there will be only one line and the font will be sized to fit up.

Because `MDCBottomNavigationBar` supports multi-line I changed this to:
```
_titleLabel.numberOfLines = 0;
```

File: components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m
Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Done.  Changed the image to a fixed height and positioning to a fixed offset from the top of the dialog.

Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationLargeItemDialogView.m]: Screen Shot 2019-04-16 at 17 30 53

**Comment by nokeeo** [comment]
> Because this is a part of a larger effort, please add context to the PR description on the need for a separate `MDCBottomNavigationLargeItemDialogView` + `MDCBottomNavigationSystemDialogView` implementation. Is it possible to consolidate these two types into one?

@jverkoey In the context of the immediate larger effort there is no need for them to be separate and they can be consolidated.  However my intent when writing this code was that that each class should have a single responsibility, `MDCBottomNavigationSystemDialogView`, laying out/styling the container view, and `MDCBottomNavigationLargeItemDialogView`, laying out/styling the content of the container.  It encapsulates each concern and increase reusability of this code if the need arises.

Any thoughts on this @romoore ?

**Comment by romoore** [comment]
@jverkoey @nokeeo Having read the code, I would say that having them separated is easy enough to read and I don't think they need to be combined. However, I can understand that since the separation isn't *necessary*, combining them can be seen as simpler. I'll let @jverkoey decide whether he means it to be a suggestion or a necessity before merging.

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (romoore) [reply components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m:76]: I think that's a fair reason. You may want to set `translatesAutoresizingMaskIntoConstraints = NO;` to make it clear you're expecting to do manual layout.

File: components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m
Comment (romoore) [inline components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m]: Nit: Please use `self.dialogView = nil;` since it's a property. Yay, properties!

File: components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m
Comment (romoore) [inline components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m:37]: Nit: Please place a blank line between the general method body and `super` calls in `tearDown` and `setUp`.

File: components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m
Comment (romoore) [inline components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m]: Tip: Place these helper methods at the top of the file (with `setUp` and `tearDown`) to keep the test methods together in a single block.

**Review by romoore — APPROVED** [review]
LGTM. Thanks for the snapshot tests! I'm not 100% excited by the "small" layout, but it's good to capture how it gets laid-out in that case (which should be never).

File: components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m
Comment (romoore) [reply components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m]: Only because manipulating the frame is undefined if there's a non-identity transform applied to the view.  It's generally safer and since this is basically just a size change (and re-centering) I thought it was easy enough to use center + bounds.

https://developer.apple.com/documentation/uikit/uiview/1622621-frame - Note the severe warning UIKit provides to manipulating `frame` when transformed.

Comment (nokeeo) [reply components/BottomNavigation/src/private/MDCBottomNavigationSystemDialogView.m]: Thanks for the explanation!

File: components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m
Comment (nokeeo) [reply components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m]: Done.

File: components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m
Comment (nokeeo) [reply components/BottomNavigation/tests/snapshot/MDCBottomNavigationLargeItemDialogViewSnapshotTests.m]: Done.

**Review by romoore — APPROVED** [review]
LGTM!

**Comment by romoore** [comment]
@jverkoey is out today. @nokeeo If you're OK with this, let me know and we can merge it in.

**Comment by nokeeo** [comment]
@romoore I am OK with the current state!

**Pull request merged**
